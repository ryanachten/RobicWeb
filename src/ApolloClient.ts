import ApolloClient from "apollo-boost";
import { API_URL } from "./constants/API";

const client = new ApolloClient({
  uri: API_URL,
  request: async operation => {
    const token = window.localStorage.getItem("token");
    operation.setContext({
      headers: {
        authorization: token
      }
    });
  }
});

export default client;
