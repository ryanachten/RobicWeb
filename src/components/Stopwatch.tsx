// Time functionality via https://codepen.io/_Billy_Brown/pen/dbJeh

import * as React from "react";
import { Typography, createStyles, Theme } from "@material-ui/core";
import { Classes } from "jss";
import { withStyles } from "@material-ui/styles";
import TimerIcon from "@material-ui/icons/Timer";

const styles = (theme: Theme) =>
  createStyles({
    icon: {
      marginRight: theme.spacing(1)
    },
    root: {
      backgroundColor: theme.palette.common.white,
      border: `1px solid ${theme.palette.text.disabled}`,
      borderRadius: "8px",
      display: "flex",
      padding: theme.spacing(1)
    }
  });

type Props = {
  classes: Classes;
  ref: any;
};

type State = {
  // laps:
  running: boolean;
  time: number | null;
  times: [number, number, number];
};

class Stopwatch extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      //   laps: [],
      running: false,
      time: null,
      times: [0, 0, 0]
    };
  }

  public componentDidMount() {
    this.reset();
  }

  public reset() {
    this.setState({
      times: [0, 0, 0]
    });
  }

  public start() {
    let { running, time } = this.state;
    if (!time) time = performance.now();
    if (!running) {
      running = true;
      requestAnimationFrame(this.step.bind(this));
    }
    this.setState({
      time,
      running
    });
  }

  // lap() {
  //   let times = this.times;
  //   li.innerText = this.format(times);
  //   this.results.appendChild(li);
  // }

  public stop() {
    this.setState({
      running: false,
      time: null
    });
  }

  public restart() {
    let { running, time } = this.state;
    if (!time) time = performance.now();
    if (!running) {
      running = true;
      requestAnimationFrame(this.step.bind(this));
    }
    this.setState({
      time,
      running
    });
    this.reset();
  }

  public step(timestamp: number) {
    if (!this.state.running) return;
    this.calculate(timestamp);
    this.setState({
      time: timestamp
    });
    requestAnimationFrame(this.step.bind(this));
  }

  public calculate(timestamp: number) {
    const { time, times } = this.state;
    if (!time) {
      return;
    }
    const diff = timestamp - time;
    // Hundredths of a second are 100 ms
    times[2] += diff / 10;
    // Seconds are 100 hundredths of a second
    if (times[2] >= 100) {
      times[1] += 1;
      times[2] -= 100;
    }
    // Minutes are 60 seconds
    if (times[1] >= 60) {
      times[0] += 1;
      times[1] -= 60;
    }
    this.setState({
      times
    });
  }

  public getTime() {
    const times = this.state.times;
    const date = new Date(0);
    date.setMinutes(times[0]);
    date.setSeconds(times[1]);
    date.setMilliseconds(times[2]);
    return date;
  }

  public render() {
    const classes = this.props.classes;
    const times = this.state.times;
    return (
      <div className={classes.root}>
        <TimerIcon className={classes.icon} color="disabled" />
        <Typography>{`${pad0(times[0], 2)}:`}</Typography>
        <Typography>{`${pad0(times[1], 2)}:`}</Typography>
        <Typography>{`${pad0(Math.floor(times[2]), 2)}`}</Typography>
      </div>
    );
  }
}

function pad0(value: number, count: number) {
  let result = value.toString();
  for (; result.length < count; --count) result = "0" + result;
  return result;
}

export default withStyles(styles)(Stopwatch);
