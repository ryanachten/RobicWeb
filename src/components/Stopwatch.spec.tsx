import React from "react";
import { createShallow } from "@material-ui/core/test-utils";
import Stopwatch from "./Stopwatch";
import renderer from "react-test-renderer";
import { ApolloProvider } from "react-apollo";
import ApolloClient from "../ApolloClient";

describe("Component: Stopwatch", () => {
  let shallow: any;

  const initProps = {
    ref: jest.fn()
  };

  beforeEach(() => {
    shallow = createShallow({ untilSelector: "Stopwatch" });
  });

  it("renders correctly", () => {
    const tree = renderer.create(<Stopwatch {...initProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should set time and running on start and stop", () => {
    const wrapper = shallow(<Stopwatch {...initProps} />);
    const instance = wrapper.instance();
    const now = 571144.6650000289;
    performance.now = jest.fn().mockImplementation(() => now);
    expect(instance.state.running).toBeFalsy;
    expect(instance.state.time).toBeNull;
    instance.start();
    expect(instance.state.running).toBeTruthy;
    expect(instance.state.time).toBe(now);
    instance.start();
    expect(instance.state.running).toBeFalsy;
    expect(instance.state.time).toBeNull;
  });

  it("should set time on reset", () => {
    const wrapper = shallow(<Stopwatch {...initProps} />);
    const instance = wrapper.instance();
    instance.state.times = [10, 10, 10];
    instance.reset();
    expect(instance.state.times).toEqual([0, 0, 0]);
  });

  it("should set time, running and call reset on restart", () => {
    const wrapper = shallow(<Stopwatch {...initProps} />);
    const instance = wrapper.instance();
    instance.state.running = true;
    instance.state.time = 571144.6650000289;
    const reset = jest.fn();
    instance.reset = reset;
    instance.restart();
    expect(instance.state.running).toBeFalsy;
    expect(instance.state.time).toBeNull;
    expect(reset).toHaveBeenCalled;
  });

  it("should calculate formatted times from timestamp", () => {
    const wrapper = shallow(<Stopwatch {...initProps} />);
    const instance = wrapper.instance();
    instance.state.time = 570000;
    instance.calculate(580000);
    expect(instance.state.times).toEqual([0, 1, 900]);
  });

  it("should set time and call calculate on step", () => {
    const wrapper = shallow(<Stopwatch {...initProps} />);
    const instance = wrapper.instance();
    const calculate = jest.fn();
    const timestamp = 570000;
    instance.calculate = calculate;
    instance.state.running = true;
    instance.step(timestamp);
    expect(calculate).toHaveBeenCalled;
    expect(instance.state.time).toBe(timestamp);
  });
});
