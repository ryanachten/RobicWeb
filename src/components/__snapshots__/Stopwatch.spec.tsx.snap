// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Stopwatch renders correctly 1`] = `
<div
  style={
    Object {
      "display": "flex",
    }
  }
>
  <p
    className="MuiTypography-root-1 MuiTypography-h5-17"
  >
    00:
  </p>
  <p
    className="MuiTypography-root-1 MuiTypography-h5-17"
  >
    00:
  </p>
  <p
    className="MuiTypography-root-1 MuiTypography-h5-17"
  >
    00
  </p>
</div>
`;
